### PR TITLE
[AutoProcessor] Correct AutoProcessor and automatically add processor…

### PR DESCRIPTION
**Pull request by patrickvonplaten**
… class.

This PR makes sure that all preprocessors, when saved automatically save the corresponding class to `preprocessing_config.json` so that they can always be loaded from `AutoProcessor` afterward. E.g. the following should always be possible:

```python
auto_processor.save_pretrained("./")
processor = AutoProcessor.from_pretrained("./")
```

### CODE REVIEW

File: src/transformers/feature_extraction_utils.py [hunk 1]
```diff
@@ -286,22 +286,24 @@ def from_pretrained(
 
         return cls.from_dict(feature_extractor_dict, **kwargs)
 
-    def save_pretrained(self, save_directory: Union[str, os.PathLike]):
+    def save_pretrained(self, save_directory: Union[str, os.PathLike], processor_class_str: Optional[str] = None):
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_utils.py:289]: this allows us to save the _processor_class_ in the feature_extractor config so that:

```python
auto_processor.save_pretrained("./")
AutoProcessor.from_pretrained("./")
```

is always possible for Wav2Vec2

File: src/transformers/models/auto/processing_auto.py [hunk 2]
```diff
@@ -162,7 +162,7 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
         model_type = config_class_to_model_type(type(config).__name__)
 
         if getattr(config, "processor_class", None) is not None:
-            processor_class = config.processor_class
+            processor_class = processor_class_from_name(config.processor_class)
```
Comment (patrickvonplaten) [inline src/transformers/models/auto/processing_auto.py:165]: This was a bug before

**Comment by patrickvonplaten** [comment]
Error is unrelated

File: src/transformers/feature_extraction_utils.py [hunk 3]
```diff
@@ -286,22 +286,24 @@ def from_pretrained(
...
         """
         Save a feature_extractor object to the directory ``save_directory``, so that it can be re-loaded using the
         :func:`~transformers.feature_extraction_utils.FeatureExtractionMixin.from_pretrained` class method.
 
         Args:
             save_directory (:obj:`str` or :obj:`os.PathLike`):
                 Directory where the feature extractor JSON file will be saved (will be created if it does not exist).
+            processor_class_str (:obj:`str`, `optional`):
+                Processor class to be saved in config.
```
Comment (sgugger) [inline src/transformers/feature_extraction_utils.py:298]: I have no idea how GitHub thinks he'll be able to merge this in the current Markdown docstring but you need a rebase and adapt that docstring here.

File: src/transformers/feature_extraction_utils.py [hunk 4]
```diff
@@ -431,16 +433,23 @@ def from_dict(cls, feature_extractor_dict: Dict[str, Any], **kwargs) -> PreTrain
...
             return feature_extractor
 
-    def to_dict(self) -> Dict[str, Any]:
+    def to_dict(self, processor_class_str: Optional[str] = None) -> Dict[str, Any]:
         """
         Serializes this instance to a Python dictionary.
 
+        Args:
+            processor_class_str (:obj:`str`, `optional`):
+                Processor class to be saved in config.
```
Comment (sgugger) [inline src/transformers/feature_extraction_utils.py:442]:
```suggestion
        Args:
            processor_class_str (`str`, *optional*):
                Processor class to be saved in config.
```
Docstrings are in Markdown now.

File: src/transformers/feature_extraction_utils.py [hunk 5]
```diff
@@ -462,32 +471,38 @@ def from_json_file(cls, json_file: Union[str, os.PathLike]) -> PreTrainedFeature
...
         return cls(**feature_extractor_dict)
 
-    def to_json_string(self) -> str:
+    def to_json_string(self, processor_class_str: Optional[str] = None) -> str:
         """
         Serializes this instance to a JSON string.
 
+        Args:
+            processor_class_str (:obj:`str`, `optional`):
+                Processor class to be saved in config.
```
Comment (sgugger) [inline src/transformers/feature_extraction_utils.py:480]: Same here

File: src/transformers/feature_extraction_utils.py [hunk 6]
```diff
@@ -462,32 +471,38 @@ def from_json_file(cls, json_file: Union[str, os.PathLike]) -> PreTrainedFeature
...
-    def to_json_file(self, json_file_path: Union[str, os.PathLike]):
+    def to_json_file(self, json_file_path: Union[str, os.PathLike], processor_class_str: Optional[str] = None):
         """
         Save this instance to a JSON file.
 
         Args:
             json_file_path (:obj:`str` or :obj:`os.PathLike`):
                 Path to the JSON file in which this feature_extractor instance's parameters will be saved.
+            processor_class_str (:obj:`str`, `optional`):
+                Processor class to be saved in config.
```
Comment (sgugger) [inline src/transformers/feature_extraction_utils.py:502]: Same here.

File: src/transformers/models/auto/processing_auto.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/models/auto/processing_auto.py:165]: Thanks for fixing!

**Review by sgugger — COMMENTED** [review]
Before being able to merge this PR, you will need to rebase and fix the docstrings. If we go with it that is.

I have a general issue with feature extractors and processors sharing the same config file. It implies every processor will always have a feature extractor but different modalities in the future might make us do processors with different components, none of them being a feature extractor.

Then why is the processor class saved in the feature extractor config and not the tokenizer config? This is all very weird and I think we should just have a separate processor config which would avoid passing the processor_class to a `sace_pretrained`  call. This feels a bit too much like a hack.

**Comment by patrickvonplaten** [comment]
> Before being able to merge this PR, you will need to rebase and fix the docstrings. If we go with it that is.
> 
> I have a general issue with feature extractors and processors sharing the same config file. It implies every processor will always have a feature extractor but different modalities in the future might make us do processors with different components, none of them being a feature extractor.
> 
> Then why is the processor class saved in the feature extractor config and not the tokenizer config? This is all very weird and I think we should just have a separate processor config which would avoid passing the processor_class to a `sace_pretrained` call. This feels a bit too much like a hack.

I understand your view and I also don't think it's very clean, but I think it's better than the alternative because:

- 1.) One [can already load the preprocessor from the feature extractor](https://github.com/huggingface/transformers/blob/13504dcbea231d2cae701d1ffdeb0810d62aff81/src/transformers/models/auto/processing_auto.py#L156) which already made me adapt a lot of feature extractor configs: https://huggingface.co/microsoft/wavlm-large/blob/main/preprocessor_config.json . If we can load it from the feature extractor, then why not save it? Reverting this behavior is already quite a breaking change by now.

- 2.) I don't think a `processor.json` config will every have any real configuration arguments. IMO it'll always stay a wrapper for tokenizers, feature extractors, language models, ... => So IMO it's really just the `"processor_class"` argument that we need to save somewhere

- 3.) We also save `tokenizer_class` in the model's config (I guess that is a bit different because `config.json` is more of an "overall" configuration. I would also be fine with having a `processor_class` in `tokenizer_config.json` as well BTW

-4.) The naming of `FeatureExtractor` is maybe general enough to use it as a base class for all future processors of whatever modality? Thinking of Video, TextToSpeech, .... 

- 5.) The problem I'm having with creating a config class for a processor is the following: 
       - Do we add the whole config of the feature extractor and the tokenizer_json.config to the processor config as well? Then we will have duplicated configs in model repos (we can't really remove existing config from repos with a huge breaking change). It'll be a mess to have duplicated configs IMO. If we don't add the tokenizer and feature extractor config then we can't load a processor just from it's config and the config will IMO only ever have `processor_class` as an argument.
 
What do you think @sgugger ? Also curious to hear other thoughts on this.

File: src/transformers/feature_extraction_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/feature_extraction_utils.py:298]: Sorry will do!

**Comment by LysandreJik** [comment]
I think the role of processors isn't well defined enough to take a decision, and I think we don't necessarily align on their purpose.

If a `Processor` is an umbrella over tokenizers and feature extractors, in that the following are possible:

```
Processor
    | Tokenizer
    | Feature extractor

Processor
    | Feature extractor

Processor
    | Tokenizer
```

Then I think that there should either be a `processor_config.json` which saves the class of the `Processor`, or that the processor class should be saved in all its dependants (tokenizer and/or feature extractor). The latter being not adding a new file (yay!) but it isn't necessarily the cleanest.

If that's not the case, and `Processor` is really an umbrella just over the feature extractor, so that only the following is possible:
```
Processor
    | Feature extractor
    | Tokenizer

Processor
    | Feature extractor
```

Then I understand why you'd want to only save it in the feature extractor configuration. But if that's the case, I wonder why there's a need to have a `Processor` at all, and why the feature extractor couldn't just contain a `tokenizer` alongside everything else that it does. If processor == feature extractor but serves as an additional abstraction so that the API is cleaner, then I also understand where you're coming from @patrickvonplaten.

**Comment by sgugger** [comment]
> 1.) One can already load the preprocessor from the feature extractor 

I don't think that's always possible. It works for Wav2Vec2-style models where the tokenizer is only used to decode the outputs, but for some model, the feature extractor **and** the tokenizer are both needed. So in that case, loading from just the feature extractor will fail.

> 2.) I don't think a processor.json config will every have any real configuration arguments. IMO it'll always stay a wrapper for tokenizers, feature extractors, language models, ... => So IMO it's really just the "processor_class" argument that we need to save somewhere

I agree with that comment, and it also does not make sense to put everything in `processor.json` since we want to be able to load the feature extractor/tokenizer separately. In this case the `processor_class` should be saved in the tokenizer config, feature extractor config and model config if accessible, just to make sure the info is widely available.

But this only works for as long as we have processors that are just grouping together some modality-specific processing components and do not need any config of their own. Maybe in the future we will have some processors that need a config? Not sure. We can certainly re-open this debate only when the case arises and try to avoid a processor config for now.

> 4.) The naming of FeatureExtractor is maybe general enough to use it as a base class for all future processors of whatever modality? Thinking of Video, TextToSpeech, ....

I think we said we would keep feature extractor modality-specific and that processors are specifically multi-modal, so we will always need processors.

So to summarize: okay for now with this PR but we also save the `processor_class` in the tokenizer config (and look for it there if we don't find a feature extractor). Note that the `save_pretrained` method of all the processors could also be abstracted in a base class if we keep the list of "processing_blocks" names (for now almost always tokenizer and feature_extractor) in a class attribute.

Last bit: to remove the hack of passing a processor_class to the `save_pretrained` methods, can we just add it as an attribute of the feature extractor/tokenizer config? It doesn't make sense for either `FeatureExtractorMixin.save_pretrained` or `PreTrainedTokenizerBase.save_pretrained` to have this argument.

**Comment by patrickvonplaten** [comment]
Added it to the tokenizer config now as well :-)

**Review by anton-l — APPROVED** [review]
Looks good! Let me know if you need an extra pair of hands for updating the old configs :)

**Comment by patrickvonplaten** [comment]
> Looks good! Let me know if you need an extra pair of hands for updating the old configs :)

Think it's fine to leave them as is :-)

File: src/transformers/feature_extraction_utils.py [hunk 7]
```diff
@@ -211,6 +211,10 @@ def __init__(self, **kwargs):
                 logger.error(f"Can't set {key} with value {value} for {self}")
                 raise err
 
+    def _set_processor_class(self, processor_class: str):
+        """Sets processor class as an attribute."""
+        self.processor_class = processor_class
```
Comment (sgugger) [inline src/transformers/feature_extraction_utils.py:216]: Don't think we should add a public attribute. Like for the tokenizer, it should be private as it's not relevant to the feature extractor, and there should be special logic to load/save it as "processor_class".

File: src/transformers/models/auto/processing_auto.py [hunk 8]
```diff
@@ -150,20 +152,28 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
...
         model_files = [f.split("/")[-1] for f in model_files]
 
+        # Let's start by checking whether the processor class is saved in a feature extractor
         if FEATURE_EXTRACTOR_NAME in model_files:
             config_dict, _ = FeatureExtractionMixin.get_feature_extractor_dict(pretrained_model_name_or_path, **kwargs)
             if "processor_class" in config_dict:
                 processor_class = processor_class_from_name(config_dict["processor_class"])
                 return processor_class.from_pretrained(pretrained_model_name_or_path, **kwargs)
 
+        # Next, let's whether the processor class is saved in a tokenizer
```
Comment (sgugger) [inline src/transformers/models/auto/processing_auto.py:162]: Great!

**Review by sgugger — APPROVED** [review]
Thanks for adapting your PR! I think one nice last step would be to have the `processor_class` be private in the feature extractor (it has nothing to do with this object) like for the tokenizer.

In any case, let's wait for @LysandreJik return and review before merging.

File: src/transformers/feature_extraction_utils.py [hunk 9]
```diff
@@ -480,6 +482,10 @@ def to_json_string(self) -> str:
         for key, value in dictionary.items():
             if isinstance(value, np.ndarray):
                 dictionary[key] = value.tolist()
+            elif key == "_processor_class" and value is not None:
+                # make sure private name "_processor_class" is correctly
+                # saved as "processor_class"
```
Comment (sgugger) [inline src/transformers/feature_extraction_utils.py]: This comment fits in one line.
(Are you nudging me toward writing a comment-styler now? ;-) )

Comment (patrickvonplaten) [reply src/transformers/feature_extraction_utils.py]: haha good catch!

**Comment by patrickvonplaten** [comment]
Merging this as it's blocking me and I think @LysandreJik is fine with the solution as well. 

@LysandreJik, please let me know if you would like to change something here before the next release?
